libkbfs/tlf_journal.go

```diff
@@ -1383,20 +1462,75 @@ func (j *tlfJournal) getBlockData(id kbfsblock.ID) (
 	return j.blockJournal.getData(id)
 }
 
+// ErrDiskLimitTimeout is returned when putBlockData exceeds
+// diskLimitTimeout when trying to acquire bytes to put.
+type ErrDiskLimitTimeout struct {
+	timeout        time.Duration
+	requestedBytes int64
+	availableBytes int64
+	err            error
+}
+
+func (e ErrDiskLimitTimeout) Error() string {
+	return fmt.Sprintf("Disk limit timeout of %s reached; requested %d bytes, only %d bytes available: %+v",
+		e.timeout, e.requestedBytes, e.availableBytes, e.err)
+}
+
 func (j *tlfJournal) putBlockData(
-	ctx context.Context, id kbfsblock.ID, context kbfsblock.Context, buf []byte,
-	serverHalf kbfscrypto.BlockCryptKeyServerHalf) error {
+	ctx context.Context, id kbfsblock.ID, blockCtx kbfsblock.Context, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) (err error) {
+	// Since Acquire can block, it should happen outside of the
+	// journal lock.
+
+	timeout := j.config.diskLimitTimeout()
+	acquireCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	bufLen := int64(len(buf))
+	j.log.CDebugf(acquireCtx, "Acquiring %d bytes for %s", bufLen, j.tlfID)
+	availableBytes, err := j.diskLimiter.Acquire(acquireCtx, bufLen)
+	switch errors.Cause(err) {
+	case nil:
+		// Continue.
+	case acquireCtx.Err():
```

Comment:
I think you need to distinguish a context canceled from a deadline exceeded here, because a timeout error wouldn't really make sense if the user hits ctrl-c.